clippy_lints/src/loops/infinite_loop.rs

```diff
@@ -128,3 +140,65 @@ impl<'hir> Visitor<'hir> for LoopVisitor<'hir, '_> {
         }
     }
 }
+
+/// Similar to [`FnRetTy`], but reveals the actual type of an `OpaqueDef`.
+enum RetTy<'hir> {
+    DefaultReturn(Span),
+    Return(Ty<'hir>),
+}
+
+impl<'hir> RetTy<'hir> {
+    fn from_fn_ret_ty(cx: &LateContext<'hir>, fn_ret_ty: FnRetTy<'hir>) -> Self {
+        /// Reveal and return the related type of an `opaque`, return `None` if the
+        /// given `ty` is not an `OpaqueDef`.
+        fn inner_<'tcx>(cx: &LateContext<'tcx>, ty: &Ty<'tcx>) -> Option<Ty<'tcx>> {
+            /// Visitor to find the type binding.
+            struct BindingVisitor<'tcx> {
+                res: Option<Ty<'tcx>>,
+            }
+            impl<'tcx> Visitor<'tcx> for BindingVisitor<'tcx> {
+                fn visit_assoc_type_binding(&mut self, type_binding: &'tcx hir::TypeBinding<'tcx>) {
+                    if self.res.is_some() {
+                        return;
+                    }
+                    if let hir::TypeBindingKind::Equality {
+                        term: hir::Term::Ty(ty),
+                    } = type_binding.kind
+                    {
+                        self.res = Some(*ty);
+                    }
+                }
+            }
+
+            let TyKind::OpaqueDef(item_id, ..) = ty.kind else {
+                return None;
+            };
+            let opaque_ty_item = cx.tcx.hir().item(item_id);
+
+            // Sinces the `item_id` is from a `TyKind::OpaqueDef`,
+            // therefore the `Item` related to it should always be `OpaqueTy`.
```

Comment:
```suggestion
            // Since `item_id` is from a `TyKind::OpaqueDef`,
            // the `Item` related to it should always be `OpaqueTy`.
```